sdk/src/main/scala/com/horizen/account/state/ForgerStakeMsgProcessor.scala

```diff
@@ -0,0 +1,709 @@
+  override val contractCode: Array[Byte] = Keccak256.hash("ForgerStakeSmartContractCode")
+
+  val networkParams: NetworkParams = params
+
+  /**
+   * TODO: Currently the stake id is calculated exclusively from message properties, which is good because it results in
+   *  a predictable result for the caller. The downside is that we need the nonce for that and in the context of RPC
+   *  calls messages often do not contain a nonce. Normal smart contract are also unable to access the nonce of an
+   *  account. A possible solution is to store a nonce counter within the forger stakes contract and use that, which
+   *  removes the need for the message nonce, but also destroys the predictability. We would need to store a nonce per
+   *  account to get that back.
+   */
+  def getStakeId(msg: Message): Array[Byte] = {
+    Keccak256.hash(Bytes.concat(
+      msg.getFrom.address(), msg.getNonce.toByteArray, msg.getValue.toByteArray, msg.getData))
+  }
+
+  override def init(view: BaseAccountStateView): Unit = {
+    super.init(view)
+    // set the initial value for the linked list last element (null hash)
+
+    // check we do not have this key set to any value yet
+    val initialTip = view.getAccountStorage(contractAddress, LinkedListTipKey)
+
+    // getting a not existing key from state DB using RAW strategy as the api is doing
+    // gives 32 bytes filled with 0 (CHUNK strategy gives an empty array instead)
+    if (!initialTip.sameElements(NULL_HEX_STRING_32))
+      throw new MessageProcessorInitializationException("initial tip already set")
+
+    view.updateAccountStorage(contractAddress, LinkedListTipKey, LinkedListNullValue)
+  }
+
+  def existsStakeData(view: BaseAccountStateView, stakeId: Array[Byte]): Boolean = {
+    // do the RAW-strategy read even if the record is actually multi-line in stateDb. It will save some gas.
+    val data = view.getAccountStorage(contractAddress, stakeId)
+    // getting a not existing key from state DB using RAW strategy
+    // gives an array of 32 bytes filled with 0, while using CHUNK strategy
+    // gives an empty array instead
+    !data.sameElements(NULL_HEX_STRING_32)
+  }
+
+  def findStakeData(view: BaseAccountStateView, stakeId: Array[Byte]): Option[ForgerStakeData] = {
+    val data = view.getAccountStorageBytes(contractAddress, stakeId)
+    if (data.length == 0) {
+      // getting a not existing key from state DB using RAW strategy
+      // gives an array of 32 bytes filled with 0, while using CHUNK strategy, as the api is doing here
+      // gives an empty array instead
+      None
+    } else {
+      ForgerStakeDataSerializer.parseBytesTry(data) match {
+        case Success(obj) => Some(obj)
+        case Failure(exception) =>
+          throw new ExecutionRevertedException("Error while parsing forger data.", exception)
+      }
+    }
+  }
+
+  def findLinkedListNode(view: BaseAccountStateView, nodeId: Array[Byte]): Option[LinkedListNode] = {
```

Comment:
The implementation of the Forger Stakes processor is getting really big and hard to read. We impove both by refactoring the linked list implementation to its own class.